heron/metricsmgr/src/java/com/twitter/heron/metricsmgr/MetricsManager.java

```diff
@@ -328,24 +332,32 @@ public void start() {
      * Handler for uncaughtException
      */
     public void uncaughtException(Thread thread, Throwable exception) {
-      String threadName = thread.getName();
       LOG.log(Level.SEVERE,
-          "Exception caught in thread: " + threadName + " with thread id: " + thread.getId(),
+          "Exception caught in thread: " + thread.getName() + " with thread id: " + thread.getId(),
           exception);
 
-      String sinkId = threadName;
-      Integer thisSinkRetryAttempts = sinksRetryAttempts.remove(sinkId);
+      if (exception instanceof Error || thread.getId() == mainThreadId) {
+        LOG.severe("Would not recover from error. Metrics Manager halts immediately");
+        Runtime.getRuntime().halt(1);
+      }
+
+      String sinkId = thread.getName();
```

Comment:
In current implementation, we enforce the thread-name the same as sink-id. Do you think we need to add an extra map to record it?